dynamiq/nodes/agents/react.py

```diff
@@ -1851,62 +1848,100 @@ def _execute_tools(self, tools_data: list[dict], config: RunnableConfig, **kwarg
-                else:
-                    tool_result = result_raw
-                    tool_files = {}
-
-                all_results.append(
-                    {
-                        "tool_name": tool_name,
-                        "success": True,
-                        "tool_input": tool_input,
-                        "result": tool_result,
-                        "files": tool_files,
-                    }
-                )
+            result_raw = self._run_tool(tool, tool_input, config, **kwargs)
+            if isinstance(result_raw, dict) and "text" in result_raw:
+                tool_result = result_raw["text"]
+                tool_files = result_raw.get("files", {})
+            else:
+                tool_result = result_raw
+                tool_files = {}
+
+            return {
+                "tool_name": tool.name,
+                "success": True,
+                "tool_input": tool_input,
+                "result": tool_result,
+                "files": tool_files,
+            }
 
-                if self.streaming.enabled and self.streaming.mode == StreamingMode.ALL:
+        def _stream_result(res: dict[str, Any]):
+            if self.streaming.enabled and self.streaming.mode == StreamingMode.ALL:
+                try:
                     self.stream_content(
-                        content={"name": tool.name, "input": tool_input, "result": tool_result},
-                        source=tool.name,
+                        content={
+                            "name": res.get("tool_name"),
+                            "input": res.get("tool_input"),
+                            "result": res.get("result"),
+                        },
+                        source=str(res.get("tool_name")),
                         step="tool",
                         config=config,
                         **kwargs,
                     )
+                except Exception as stream_err:
+                    logger.error(f"Streaming error for tool {res.get('tool_name')}: {stream_err}")
+
+        max_workers = max(1, len(tools_data))
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            future_map = {}
+            for idx, td in enumerate(tools_data):
+                tool_name = td.get("name")
+                tool_input = td.get("input")
+                if tool_name is None or tool_input is None:
+                    error_message = "Invalid tool payload: missing 'name' or 'input'"
+                    logger.error(error_message)
+                    all_results.append(
+                        {
+                            "order": idx,
+                            "tool_name": tool_name or UNKNOWN_TOOL_NAME,
+                            "success": False,
+                            "tool_input": tool_input,
+                            "result": error_message,
+                            "files": {},
+                        }
+                    )
+                    continue
```

Comment:
Please prepare this data for parrallel running before ThreadPoolExecutor init. Inside just go over list/dict and submit function with args/kwargs.
It would look much cleaner and could help to avoid unnecessary threadpool creation in case if here just 0-1 tools (check before pool creation)